File: 303.range-sum-query-immutable.py
#
# @lc app=leetcode id=303 lang=python3
#
# [303] Range Sum Query - Immutable
#

# Time:  O(1), ctor: O(n)
# Space: O(n)

from typing import List


# @lc code=start
class NumArray:
    def __init__(self, nums: List[int]):
        self.dp = [0]
        for num in nums:
            self.dp.append(self.dp[-1] + num)

    def sumRange(self, left: int, right: int) -> int:
        return self.dp[right + 1] - self.dp[left]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)
# @lc code=end

File: 303.range-sum-query-immutable.go
/*
 * @lc app=leetcode id=303 lang=golang
 *
 * [303] Range Sum Query - Immutable
 */

package leetcode

// @lc code=start
type NumArray struct {
	dp []int
}

func Constructor(nums []int) NumArray {
	for i := 1; i < len(nums); i++ {
		nums[i] += nums[i-1]
	}
	return NumArray{dp: nums}
}

func (this *NumArray) SumRange(left int, right int) int {
	if left > 0 {
		return this.dp[right] - this.dp[left-1]
	}
	return this.dp[right]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(left,right);
 */
// @lc code=end
